### ISSUE TITLE
OnlyOffice-Nextcloud Connector not working in Nextcloud 30.0 (Connecting Refused)

### ISSUE THREAD

**Issue by EmilSepa** [ONLYOFFICE/onlyoffice-nextcloud#1020]
Hey OnlyOffice-Team, 

first of all, thanks for your great work. I really apreciate it.
Due to a bug in Nextcloud 29.0x, I unfortunately had to update to Nextcloud 30.0 early. Since then, the OnlyOffice Connector no longer works properly.
When saving the settings in the admin settings, I am told that the connection is working. When I test the connection from Nextcloud to OnlyOffice via Console, I am also shown a successful connection.
However, I always get the following message when opening/creating a file: 

“OnlyOffice not reachable. Please contact an admin”

In the Nginx logs of OnlyOffice I get the following message:

ERRCON Connection Refused 

I use nextcloud and onlyoffice behind a proxy. The problem did not disapear when using an ip-address for the connector.

Before the update the OnlyOffice worked great. I already tried older OnlyOffice Versions, but no success.

**Comment by Mer0me** [comment]
I get this error in the web console when an office document is opened : 
```
Content-Security-Policy : Les paramètres de la page ont empêché l’exécution d’un script (script-src-elem) à l’adresse https://office.mycloud.net/web-apps/apps/api/documents/api.js car il enfreint la directive suivante : « script-src-elem 'strict-dynamic' 'nonce-lX1CU4KYMEKziubGpiKZrZ9fZNlnXFET9pqqmhhKayA=' blob: https://office.mycloud.net/ » [405241](https://cloud.mycloud.net/index.php/apps/onlyoffice/405241?filePath=%2Ftest.xlsx)
Error: Could not establish connection. Receiving end does not exist.
```

**Comment by Mer0me** [comment]
Workaround : 
On /var/www/nextcloud/lib/public/AppFramework/Http/ContentSecurityPolicy.php at line 34, replace 

`protected $strictDynamicAllowedOnScripts = true; `

by

`protected $strictDynamicAllowedOnScripts = false; `

With this, OnlyOffice connector is working again in Nextcloud 30.

**Comment by Bad-and-Mad** [comment]
Same here. The workaroud 

`protected $strictDynamicAllowedOnScripts = false;`

helped me out.

**Comment by EmilSepa** [comment]
@Mer0me thanks for the workaround. Sadly, I can't find the file. I'm using the linuxserver container.

**Comment by eXist06** [comment]
> @Mer0me thanks for the workaround. Sadly, I can't find the file. I'm using the linuxserver container.

/var/www/html/lib/public/AppFramework/Http check this location

**Comment by Bad-and-Mad** [comment]
> @Mer0me thanks for the workaround. Sadly, I can't find the file. I'm using the linuxserver container.

`find / -type f -name ContentSecurityPolicy.php`

**Comment by c-prompt** [comment]
Potentially related: [NC 29/30 + OO ONLYOFFICE cannot be reached. Please contact admin](https://help.nextcloud.com/t/nc-29-30-oo-onlyoffice-cannot-be-reached-please-contact-admin/201524).

I added the following to my .htaccess per [this comment](https://help.nextcloud.com/t/nc-29-30-oo-onlyoffice-cannot-be-reached-please-contact-admin/201524/9) and OO now works properly:
```
Header onsuccess unset Content-Security-Policy
    Header always set Content-Security-Policy "frame-ancestors https://cloud.blah.com”
    RequestHeader set X-Forwarded-Proto https
    Header always set Strict-Transport-Security "max-age=15552000; includeSubDomains"
```

**Comment by edgeofthex** [comment]
Howdy All,

If someone knows the appropriate entries to add to an nginx.conf, I would surely appreciate it if they would post it.

**Comment by Bad-and-Mad** [comment]
> Howdy All,
> 
> If someone knows the appropriate entries to add to an nginx.conf, I would surely appreciate it if they would post it.

```
add_header Content-Security-Policy "";	
add_header Content-Security-Policy "frame-ancestors 'https://cloud.blah.com';" always;
proxy_set_header X-Forwarded-Port 443;
proxy_set_header X-Forwarded-Proto https;
add_header Strict-Transport-Security “max-age=15552000; includeSubDomains” always;
```

**Comment by nemphys** [comment]
Any chance this can be fixed soon? The web server (nginx) workaround does not seem to work and, as stated above, changing ContentSecurityPolicy.php breaks the mail app (along with other stuff I suppose).

**Comment by edgeofthex** [comment]
> > Howdy All,
> > If someone knows the appropriate entries to add to an nginx.conf, I would surely appreciate it if they would post it.
> 
> ```
> add_header Content-Security-Policy "";	
> add_header Content-Security-Policy "frame-ancestors 'https://cloud.blah.com';" always;
> proxy_set_header X-Forwarded-Port 443;
> proxy_set_header X-Forwarded-Proto https;
> add_header Strict-Transport-Security “max-age=15552000; includeSubDomains” always;
> ```

Thanks for the reply! That didn't work with my configuration, the OO connector refused to connect after adding the above, and modifying my hostname.  I do suppose the connector will be fixed soon.

**Comment by razerrazer** [comment]
This is not working for nginx i can confirm this as well.

**Comment by gaussb** [comment]
The workaround patching `ContentSecurityPolicy.php` also breaks  notes app

**Comment by Bad-and-Mad** [comment]
In my case, the problem was not the notes app but the text app. You should try what happens if you deactivate the text app.

**Comment by Bad-and-Mad** [comment]
> The workaround patching `ContentSecurityPolicy.php` also breaks notes app

In my case, the problem was not the notes app but the text app. You should try what happens if you deactivate the text app.

**Comment by Bad-and-Mad** [comment]
I think that changing the `ContentSecurityPolicy.php` file is not a sensible solution. It has too many effects on other apps. When using apache, the easiest way is to change the `.htaccess` file as mentioned above. In addition, with this solution you don't have to tinker with the Nextcloud program code, which is not such a good idea in principle.

Here once again the adjustments for Apache (`.htaccess` file in nextcloud route-folder):

```
<IfModule mod_env.c>
    Header onsuccess unset Content-Security-Policy
    Header always set Content-Security-Policy "frame-ancestors https://yourhost.xyz"
    RequestHeader set X-Forwarded-Proto https
    Header always set Strict-Transport-Security "max-age=15552000; includeSubDomains"
</IfModule>
```

**Comment by SpencerMackey** [comment]
@Bad-and-Mad, I appreciate your help with this, even as you are not running nginx yourself. I would agree that the adjustments for apache and nginx look equivalent.

Adding the five lines to the nginx conf gave me initially a new error. Firefox refused to open the page in a frame. Trying to open it in a new tab pushed me back to the original error message. Patching ContentSecurityPolicy.php fixes the onlyoffice issue, but breaks various other things indeed. Anyone with another idea for nginx?

**Comment by sebarnd** [comment]
I think a found the problem and a solution: 

the file https://github.com/ONLYOFFICE/onlyoffice-nextcloud/blob/master/templates/editor.php contains in line 55:
´´´
        <script nonce="<?php p(base64_encode($_["requesttoken"])) ?>"
´´´
changing it to
´´´
        <script nonce="<?php p($_["cspNonce"]) ?>"
´´´

solved the problem for me...

**Comment by Bad-and-Mad** [comment]
> I think a found the problem and a solution:
> 
> the file https://github.com/ONLYOFFICE/onlyoffice-nextcloud/blob/master/templates/editor.php contains in line 55:
> 
> ```
>         <script nonce="<?php p(base64_encode($_["requesttoken"])) ?>"
> ```
> 
> changing it to
> 
> ```
>         <script nonce="<?php p($_["cspNonce"]) ?>"
> ```
> 
> solved the problem for me...

I have applied the change in my installation and the solution works. So the changes in the `.htaccess `or the `ContentSecurityPolicy.php` are no longer necessary.

Thanks for that.

**Comment by nemphys** [comment]
True, this seems to be a proper fix and it only affects the OO plugin (as it should).

**Comment by QurQuma** [comment]
where modified 

> I think a found the problem and a solution:
> 
> the file https://github.com/ONLYOFFICE/onlyoffice-nextcloud/blob/master/templates/editor.php contains in line 55:
> 
> ```
>         <script nonce="<?php p(base64_encode($_["requesttoken"])) ?>"
> ```
> 
> changing it to
> 
> ```
>         <script nonce="<?php p($_["cspNonce"]) ?>"
> ```
> 
> solved the problem for me...

with this I receive this error:

Undefined array key "cspNonce" at /var/www/next/apps/onlyoffice/templates/editor.php#55

what’s going wrong?

**Comment by sebarnd** [comment]
> with this I receive this error:
> 
> Undefined array key "cspNonce" at /var/www/next/apps/onlyoffice/templates/editor.php#55
> 
> what’s going wrong?
Maybe you're using a nextcloud version smaller v30.0.0rc1? The `cspNonce` key was introduced quite recently see also https://github.com/nextcloud/server/commit/2916e5df7e08fc588e752beaf486d907112a34ee

**Comment by LinneyS** [comment]
Resolved at 3bfa0539882be6c1758247a04f9ff1ade9f55b1e

**Comment by marco-calautti** [comment]
In the meanwhile for those of you having the issue, and are using traefik as the reverse proxy, I fixed the issue by adding the following label to the *Nextcloud* entry of my docker compose file:

`traefik.http.middlewares.nextcloud-onlyoffice-policy.headers.customResponseHeaders.Content-Security-Policy: frame-ancestors 'self';`

Remember to add `nextcloud-onlyoffice-policy` to the list of middlewares.
